File: python/p_0005.py
#!/usr/bin/env python3

"""What is the smallest positive number that is evenly divisible by
 all of the numbers from 1 to 20?"""

from functools import reduce

def gcd(a: int, b: int) -> int:
  """Calculates the greatest common divisor (GCD) of two integers."""
  if b == 0:
    return a
  else:
    return gcd(b, a % b)

def lcm(a: int, b: int) -> int:
  """Calculates the least common multiple (LCM) of two integers."""
  return a * b // gcd(a, b)

if __name__ == "__main__":
  result = reduce(lcm, range(1, 21))
  print(result)

File: cpp/src/p_0005.cc
// What is the smallest positive number that is evenly divisible by all of the
// numbers from 1 to 20?

#include "euler.h"
#include "iota_view.h"

uint64_t gcd(uint64_t a, uint64_t b) { return (b == 0) ? a : gcd(b, a % b); }

auto lcm(uint64_t a, uint64_t b) { return a * b / gcd(a, b); }

uint64_t euler_0005(int max) {
    auto numbers = iota_view{2, max + 1};
    return std::accumulate(
        numbers.begin(), numbers.end(), 1,
        [](uint64_t result, int x) { return lcm(result, x); });
}
